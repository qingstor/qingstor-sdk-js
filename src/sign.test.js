// +-------------------------------------------------------------------------
// | Copyright (C) 2016 Yunify, Inc.
// +-------------------------------------------------------------------------
// | Licensed under the Apache License, Version 2.0 (the "License");
// | you may not use this work except in compliance with the License.
// | You may obtain a copy of the License in the LICENSE file, or at:
// |
// | http://www.apache.org/licenses/LICENSE-2.0
// |
// | Unless required by applicable law or agreed to in writing, software
// | distributed under the License is distributed on an "AS IS" BASIS,
// | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// | See the License for the specific language governing permissions and
// | limitations under the License.
// +-------------------------------------------------------------------------

import process from 'process';
import util from 'util';
import Signer from './sign';
const packageVersion = require('../package.json').version;

const userAgent = util.format(
  'qingstor-sdk-js/%s (Node.js %s; %s %s)',
  packageVersion,
  process.version,
  process.platform,
  process.arch
);

const operation = {
  endpoint: 'https://pek3a.qingstor.com:443',
  method: 'PUT',
  path: '/test_bucket/test_object.jpg?acl',
  params: {
    upload_id: 'test_upload_id',
  },
  uri: 'https://pek3a.qingstor.com:443/test_bucket/test_object.jpg?acl&upload_id=test_upload_id',
  body: 'test string',
  headers: {
    host: 'qingstor.com',
    'x-qs-date': 'Fri, 01 Nov 2019 07:40:08 GMT',
    'content-length': 11,
    'content-type': 'image/jpeg',
    'x-qs-z': 'test-z',
    'x-qs-a': 'test-a',
    'x-qs-a-abc': 'test-abc',
    'user-agent': userAgent,
  },
};

const signer = new Signer('test_key', 'test_secret');

test('sign on header', () => {
  const signedOperation = signer.sign(operation);
  expect(signedOperation).toMatchSnapshot({
    headers: {
      'user-agent': expect.stringContaining('qingstor-sdk-js'),
    },
  });
});

test('sign on uri', () => {
  const signedOperation = signer.signQuery({
    expires: 1572592959,
    ...operation,
  });
  expect(signedOperation).toMatchSnapshot();
});

test('getSignature', () => {
  const signature = signer.getSignature(operation);
  expect(signature).toMatchSnapshot({
    signed_date: expect.any(String),
  });
});

test('getQuerySignature', () => {
  const signature = signer.getQuerySignature({
    ...operation,
    expiresTTL: 300,
  });
  expect(signature).toMatchSnapshot({
    expires: expect.any(Number),
    signature: expect.any(String),
    signed_date: expect.any(String),
  });
});

test('getCanonicalizedResource', () => {
  signer._setOperation(operation);
  expect(signer.getCanonicalizedResource()).toMatchSnapshot();
});

test('getCanonicalizedResource by virtual host', () => {
  const _operation = {
    ...operation,
    path: '/test_object.jpg?acl',
    endpoint: 'https://test_bucket.pek3a.qingstor.com',
    uri: 'https://test_bucket.pek3a.qingstor.com/test_object.jpg?acl&upload_id=test_upload_id',
  };
  const vHostSigner = new Signer('test_key', 'test_secret', true);

  vHostSigner._setOperation(_operation);
  expect(vHostSigner.getCanonicalizedResource()).toMatchSnapshot();
});

test('getAuthorization', () => {
  signer._setOperation(operation);
  expect(signer.getAuthorization()).toMatchSnapshot();
});
