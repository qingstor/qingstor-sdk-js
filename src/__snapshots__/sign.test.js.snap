// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAuthorization 1`] = `"L6q5iKoyWbQky0T5/MYELz5jAFecGbvLDyjLdz6fClw="`;

exports[`getCanonicalizedResource 1`] = `"/test_bucket/test_object.jpg?acl&upload_id=test_upload_id"`;

exports[`getCanonicalizedResource by virtual host 1`] = `"/test_bucket/test_object.jpg?acl&upload_id=test_upload_id"`;

exports[`getQuerySignature 1`] = `
Object {
  "access_key_id": "test_key",
  "expires": Any<Number>,
  "signature": Any<String>,
  "signed_date": Any<String>,
}
`;

exports[`getSignature 1`] = `
Object {
  "authorization": "QS test_key:L6q5iKoyWbQky0T5/MYELz5jAFecGbvLDyjLdz6fClw=",
  "signed_date": Any<String>,
}
`;

exports[`sign on header 1`] = `
Object {
  "body": "test string",
  "endpoint": "https://pek3a.qingstor.com:443",
  "headers": Object {
    "authorization": "QS test_key:ZwavbCq3ylBMlU3fb5EFOabtNyUYWzqi2e0rQBTHkcM=",
    "content-length": 11,
    "content-type": "image/jpeg",
    "host": "qingstor.com",
    "user-agent": StringContaining "qingstor-sdk-js",
    "x-qs-a": "test-a",
    "x-qs-a-abc": "test-abc",
    "x-qs-date": "Fri, 01 Nov 2019 07:40:08 GMT",
    "x-qs-z": "test-z",
  },
  "method": "PUT",
  "params": Object {
    "upload_id": "test_upload_id",
  },
  "path": "/test_bucket/test_object.jpg?acl",
  "uri": "https://pek3a.qingstor.com:443/test_bucket/test_object.jpg?acl&upload_id=test_upload_id",
}
`;

exports[`sign on uri 1`] = `
Object {
  "body": "test string",
  "endpoint": "https://pek3a.qingstor.com:443",
  "expires": 1572592959,
  "headers": Object {
    "authorization": "QS test_key:ZwavbCq3ylBMlU3fb5EFOabtNyUYWzqi2e0rQBTHkcM=",
    "x-qs-a": "test-a",
    "x-qs-a-abc": "test-abc",
    "x-qs-z": "test-z",
  },
  "method": "PUT",
  "params": Object {
    "access_key_id": "test_key",
    "expires": 1572592959,
    "signature": "HjroQB6ac19fmFt/1rOMHMco0eJ17gGAGexLYD98u88=",
    "upload_id": "test_upload_id",
  },
  "path": "/test_bucket/test_object.jpg?acl",
  "uri": "https://pek3a.qingstor.com:443/test_bucket/test_object.jpg?acl&upload_id=test_upload_id&signature=HjroQB6ac19fmFt%2F1rOMHMco0eJ17gGAGexLYD98u88%3D&access_key_id=test_key&expires=1572592959",
}
`;
